beacon-chain/rpc/eth/debug/structs.go

```diff
@@ -0,0 +1,965 @@
+	if err != nil {
+		return nil, err
+	}
+	srcbr := st.BlockRoots()
+	br := make([]string, len(srcbr))
+	for i, r := range srcbr {
+		br[i] = hexutil.Encode(r)
+	}
+	srcsr := st.StateRoots()
+	sr := make([]string, len(srcsr))
+	for i, r := range srcsr {
+		sr[i] = hexutil.Encode(r)
+	}
+	srchr, err := st.HistoricalRoots()
+	if err != nil {
+		return nil, err
+	}
+	hr := make([]string, len(srchr))
+	for i, r := range srchr {
+		hr[i] = hexutil.Encode(r)
+	}
+	e1d, err := shared.Eth1DataFromConsensus(st.Eth1Data())
+	if err != nil {
+		return nil, err
+	}
+	srcvotes := st.Eth1DataVotes()
+	votes := make([]*shared.Eth1Data, len(srcvotes))
+	for i, e := range srcvotes {
+		votes[i], err = shared.Eth1DataFromConsensus(e)
+		if err != nil {
+			return nil, err
+		}
+	}
+	srcvals := st.Validators()
+	vals := make([]*Validator, len(srcvals))
+	for i, v := range srcvals {
+		vals[i], err = ValidatorFromConsensus(v)
+		if err != nil {
+			return nil, err
+		}
+	}
+	srcbals := st.Balances()
+	bals := make([]string, len(srcbals))
+	for i, b := range srcbals {
+		bals[i] = fmt.Sprintf("%d", b)
+	}
+	srcrm := st.RandaoMixes()
+	rm := make([]string, len(srcrm))
+	for i, m := range srcrm {
+		rm[i] = hexutil.Encode(m)
+	}
+	srcslashings := st.Slashings()
+	slashings := make([]string, len(srcslashings))
+	for i, s := range srcslashings {
+		slashings[i] = fmt.Sprintf("%d", s)
+	}
+	srcPrevAtts, err := st.PreviousEpochAttestations()
+	if err != nil {
+		return nil, err
+	}
+	prevAtts := make([]*PendingAttestation, len(srcPrevAtts))
+	for i, a := range srcPrevAtts {
+		prevAtts[i], err = PendingAttestationFromConsensus(a)
+		if err != nil {
+			return nil, err
+		}
+	}
+	srcCurrAtts, err := st.CurrentEpochAttestations()
+	if err != nil {
+		return nil, err
+	}
+	currAtts := make([]*PendingAttestation, len(srcCurrAtts))
+	for i, a := range srcCurrAtts {
+		currAtts[i], err = PendingAttestationFromConsensus(a)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &BeaconState{
+		GenesisTime:                 fmt.Sprintf("%d", st.GenesisTime()),
+		GenesisValidatorsRoot:       hexutil.Encode(st.GenesisValidatorsRoot()),
+		Slot:                        fmt.Sprintf("%d", st.Slot()),
+		Fork:                        f,
+		LatestBlockHeader:           shared.BeaconBlockHeaderFromConsensus(st.LatestBlockHeader()),
+		BlockRoots:                  br,
+		StateRoots:                  sr,
+		HistoricalRoots:             hr,
+		Eth1Data:                    e1d,
+		Eth1DataVotes:               votes,
+		Eth1DepositIndex:            fmt.Sprintf("%d", st.Eth1DepositIndex()),
+		Validators:                  vals,
+		Balances:                    bals,
+		RandaoMixes:                 rm,
+		Slashings:                   slashings,
+		PreviousEpochAttestations:   prevAtts,
+		CurrentEpochAttestations:    currAtts,
+		JustificationBits:           hexutil.Encode(st.JustificationBits()),
+		PreviousJustifiedCheckpoint: shared.CheckpointFromConsensus(st.PreviousJustifiedCheckpoint()),
```

Comment:
we currently don't return any error for this if it's nil, do you think we should add a check for this?

Reply:
Indeed, `FromConsensus` functions don't all follow the same pattern. I added a new checkpoint to https://github.com/prysmaticlabs/prysm/issues/12958. On the other hand, I claim that objects coming from core code will always be valid, so these nil checks feel overly defensive.